### PR TITLE
longjmp -> throw

### PR DESCRIPTION
**Pull request by jfbastien**
Older tsan doesn't like longjmp. Use throw.

### CODE REVIEW

**Comment by jfbastien** [comment]
Not quite good to go yet...

**Comment by kripken** [comment]
Works for me locally. If lgty then feel free to squash into yours and merge.

File: src/binaryen-shell.cpp [hunk 1]
```diff
@@ -275,20 +272,22 @@ static void run_asserts(size_t* i, bool* checked, AllocatingModule* wasm,
...
         std::cerr << "printing in module invalidity test:\n";
         Colors::normal(std::cout);
         std::cout << wasm;
       }
+      bool invalid = false;
+      try {
+        *builder = std::unique_ptr<SExpressionWasmBuilder>(
+            new SExpressionWasmBuilder(wasm, *curr[1], [&]() {
+              invalid = true;
+              throw TrapException();
```
Comment (kripken) [inline src/binaryen-shell.cpp]: hmm actually this and the catch below should be for a third type, a `ParseException` perhaps? It's not a runtime trap.

File: src/binaryen-shell.cpp [hunk 2]
```diff
@@ -275,20 +272,22 @@ static void run_asserts(size_t* i, bool* checked, AllocatingModule* wasm,
...
+            }));
+      } catch (const TrapException& e) {
+        invalid = true;
```
Comment (kripken) [inline src/binaryen-shell.cpp]: and I guess I added this line and it's not strictly necessary. but maybe nice for clarity, better than the setting 4 lines above?

**Comment by jfbastien** [comment]
Updated and squashed.

**Comment by jfbastien** [comment]
Odd, the following three tests now seem to reliably fail with tsan on:
	20040811-1.c.s.wast
	pr43220.c.s.wast
	vla-dealloc-1.c.s.wast
But pass on other configurations.

**Comment by kripken** [comment]
Did you get that info from the current build logs? I'm not sure what to read there, I don't see TSAN errors but just it looks like it halts?

**Comment by jfbastien** [comment]
Right, the three tests just print out FAILED and nothing else. I can't repro the problem by just running one of the tests. It's only in the tsan builds though!

**Comment by kripken** [comment]
Weird. Should we just merge this? It's obviously better than the current code.

**Comment by jfbastien** [comment]
Sure, let's merge and move tsan to optional while it stays broken (and try to fix it).

**Pull request merged**